server/etcdserver/api/v3discovery/discovery.go

```diff
@@ -0,0 +1,542 @@
+}
+
+type memberInfo struct {
+	peerURL   string
+	createRev int64
+}
+
+// key prefix for each cluster: "/_etcd/registry/<ClassId>".
+func geClusterKeyPrefix(cluster string) string {
+	return path.Join(discoveryPrefix, cluster)
+}
+
+// key format for cluster size: "/_etcd/registry/<ClassId>/_config/size".
+func geClusterSizeKey(cluster string) string {
+	return path.Join(geClusterKeyPrefix(cluster), "_config/size")
+}
+
+// key prefix for each member: "/_etcd/registry/<ClassId>/members".
+func getMemberKeyPrefix(cluster string) string {
+	return path.Join(geClusterKeyPrefix(cluster), "members")
+}
+
+// key format for each member: "/_etcd/registry/<ClassId>/members/<memberId>".
+func getMemberKey(cluster, memberId string) string {
+	return path.Join(getMemberKeyPrefix(cluster), memberId)
+}
+
+// GetCluster will connect to the discovery service at the given url and
+// retrieve a string describing the cluster
+func GetCluster(lg *zap.Logger, cfg *DiscoveryConfig) (string, error) {
+	d, err := newDiscovery(lg, cfg, 0)
+	if err != nil {
+		return "", err
+	}
+	defer d.close()
+	return d.getCluster()
+}
+
+// JoinCluster will connect to the discovery service at the given url, and
+// register the server represented by the given id and config to the cluster
+func JoinCluster(lg *zap.Logger, cfg *DiscoveryConfig, id types.ID, config string) (string, error) {
+	d, err := newDiscovery(lg, cfg, id)
+	if err != nil {
+		return "", err
+	}
+	defer d.close()
+	return d.joinCluster(config)
+}
+
+type discovery struct {
+	lg      *zap.Logger
+	cluster string
+	id      types.ID
+	c       *clientv3.Client
+	retries uint
+	durl    string
+
+	cfg *DiscoveryConfig
+
+	clock clockwork.Clock
+}
+
+func newDiscovery(lg *zap.Logger, dcfg *DiscoveryConfig, id types.ID) (*discovery, error) {
+	if lg == nil {
+		lg = zap.NewNop()
+	}
+	u, err := url.Parse(dcfg.Url)
+	if err != nil {
+		return nil, err
+	}
+	token := u.Path
+	u.Path = ""
+
+	cfg, err := newClientCfg(dcfg, u)
+	if err != nil {
+		return nil, err
+	}
+
+	c, err := clientv3.New(*cfg)
+	if err != nil {
+		return nil, err
+	}
+	return &discovery{
+		lg:      lg,
+		cluster: token,
+		id:      id,
+		c:       c,
+		durl:    u.String(),
+		cfg:     dcfg,
+		clock:   clockwork.NewRealClock(),
+	}, nil
+}
+
+// The following function follows the same logic as etcdctl, refer to
+// https://github.com/etcd-io/etcd/blob/f9a8c49c695b098d66a07948666664ea10d01a82/etcdctl/ctlv3/command/global.go#L191-L250
+func newClientCfg(dcfg *DiscoveryConfig, dUrl *url.URL) (*clientv3.Config, error) {
+	var cfgtls *transport.TLSInfo
+
+	if dcfg.CertFile != "" || dcfg.KeyFile != "" || dcfg.TrustedCAFile != "" || dcfg.ServiceName != "" {
+		cfgtls = &transport.TLSInfo{
+			CertFile:      dcfg.CertFile,
+			KeyFile:       dcfg.KeyFile,
+			TrustedCAFile: dcfg.TrustedCAFile,
+			ServerName:    dcfg.ServiceName,
+		}
+		if dcfg.InsecureDiscovery {
+			cfgtls.ServerName = ""
+		}
+		cfgtls.Logger, _ = zap.NewProduction()
```

Comment:
I think we should avoid constructing logger here. Could it maybe be passed down from `newDiscovery` function?

Reply:
good point. Will fix .

Reply:
Fixed & thanks.